prog/rand.go

```diff
@@ -582,8 +583,9 @@ func (r *randGen) generateParticularCall(s *state, meta *Syscall) (calls []*Call
 	}
 	c := MakeCall(meta, nil)
 	c.Args, calls = r.generateArgs(s, meta.Args, DirIn)
+	moreCalls, _ := r.patchConditionalFields(c, s)
```

Comment:
What are the rules for conditional fields vs len type?
If we call patchConditionalFields before assignSizesCall, I assume sizes are not assigned when we evaluate conditional fields, does it mean that len fields cannot be used in conditional field expressions?
Similarly, does len of a struct that contains conditional fields work?

Reply:
Conditional fields may only reference `typeInt` fields in their expressions, so there's no need to set sizes before we call `patchConditionalFields()`. But sizes do depend on them, so they should only be calculated after we have ensured that conditional fields are set correctly.

> Similarly, does len of a struct that contains conditional fields work?

If `len` works fine with structs that contain `optional[T]`, then it should also work with conditional fields -- the underlying implementation is similar. I can add some tests just in case.

Reply:
> Conditional fields may only reference typeInt fields in their expressions

I see.

> If len works fine with structs that contain optional[T], then it should also work with conditional fields

But if sizes are calculated before patchConditionalFields, does it mean that size will always include the field even if it will be later removed?

Reply:
Before `patchConditionalFields()` is called, some conditional fields may be inconsistent (e.g. because we mutated some int field) or may not yet have been calculated at all (e.g. because we've just generated the type). If we calculate sizes on such a state, the assigned sizes will likely be just incorrect.

Reply:
So what is the size of such structs? Is it random?
Let's capture this in the docs.

Reply:
I wouldn't say it's random -- it'd be a bug in syzkaller code if lens are not updated after conditional fields are patched.

I've added a few `size_test.go` tests.